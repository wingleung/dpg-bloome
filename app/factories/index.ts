import type { Item } from '../items'

type ItemFactory = (item: Item) => Item

export type { ItemFactory }

export * from './default-item'
