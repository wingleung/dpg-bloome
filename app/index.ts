export * from './items'
export * from './shop'
