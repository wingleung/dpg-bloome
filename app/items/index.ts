export * from './item'
export * from './aged-brie'
export * from './back-stage'
export * from './conjured'
export * from './sulfuras'

export const ItemTypes = {
  AGED_BRIE: 'Aged Brie',
  BACKSTAGE_PASS: 'Backstage passes to a TAFKAL80ETC concert',
  CONJURED: 'Conjured Mana Cake',
  SULFURAS: 'Sulfuras, Hand of Ragnaros',
} as const
