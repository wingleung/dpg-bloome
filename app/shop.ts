import type { ItemFactory } from './factories'
import type { Item } from './items'

import { defaultItemFactory } from './factories'

class Shop {
  constructor(
    public items: Item[] = [],
    private readonly itemFactory: ItemFactory = defaultItemFactory
  ) {
    // So in reality, I would talk to the goblin and check with him if I can change the Items property.
    // For the purposes of this exercise, I talked to the goblin in my head and he said it's fine :-)

    this.items = items.map(this.itemFactory)
  }

  updateQuality() {
    for (const item of this.items) {
      item.updateQuality?.()
    }

    return this.items
  }
}

export { Shop }
