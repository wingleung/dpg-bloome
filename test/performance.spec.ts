import { describe, expect, it } from 'vitest'

import { Item, Shop } from '../app'

const PERFORMANCE_THRESHOLD = 1

function benchmark(fn: () => void, iterations = 1000): number {
  const start = performance.now()
  for (let i = 0; i < iterations; i++) {
    fn()
  }
  const end = performance.now()
  return (end - start) / iterations // Average time per operation in milliseconds
}

describe('Performance Tests', () => {
  describe('empty state', () => {
    it('should benchmark empty shop performance', () => {
      const gildedRose = new Shop()

      const avgTime = benchmark(() => {
        gildedRose.updateQuality()
      })

      expect(avgTime).toBeLessThan(PERFORMANCE_THRESHOLD)
      expect({
        name: 'Empty shop',
        avgTime,
        withinThreshold: avgTime < PERFORMANCE_THRESHOLD,
      }).toMatchSnapshot({ avgTime: expect.any(Number) })
    })
  })

  describe('default item', () => {
    it('should benchmark default item performance', () => {
      const defaultItem = new Item('foo', 1, 2)
      const gildedRose = new Shop([defaultItem])

      const avgTime = benchmark(() => {
        gildedRose.updateQuality()
      })

      expect(avgTime).toBeLessThan(PERFORMANCE_THRESHOLD)
      expect({
        name: 'Default item',
        avgTime,
        withinThreshold: avgTime < PERFORMANCE_THRESHOLD,
      }).toMatchSnapshot({ avgTime: expect.any(Number) })
    })
  })

  describe('Sulfuras performance', () => {
    it('should benchmark Sulfuras performance', () => {
      const sulfurasItem = new Item('Sulfuras, Hand of Ragnaros', 0, 80)
      const gildedRose = new Shop([sulfurasItem])

      const avgTime = benchmark(() => {
        gildedRose.updateQuality()
      })

      expect(avgTime).toBeLessThan(PERFORMANCE_THRESHOLD)
      expect({
        name: 'Sulfuras',
        avgTime,
        withinThreshold: avgTime < PERFORMANCE_THRESHOLD,
      }).toMatchSnapshot({ avgTime: expect.any(Number) })
    })
  })

  describe('Aged Brie performance', () => {
    it('should benchmark Aged Brie performance', () => {
      const agedBrieItem = new Item('Aged Brie', 1, 0)
      const gildedRose = new Shop([agedBrieItem])

      const avgTime = benchmark(() => {
        gildedRose.updateQuality()
      })

      expect(avgTime).toBeLessThan(PERFORMANCE_THRESHOLD)
      expect({
        name: 'Aged Brie',
        avgTime,
        withinThreshold: avgTime < PERFORMANCE_THRESHOLD,
      }).toMatchSnapshot({ avgTime: expect.any(Number) })
    })
  })

  describe('Backstage passes performance', () => {
    it('should benchmark Backstage passes performance', () => {
      const backStageItem = new Item('Backstage passes to a TAFKAL80ETC concert', 11, 0)
      const gildedRose = new Shop([backStageItem])

      const avgTime = benchmark(() => {
        gildedRose.updateQuality()
      })

      expect(avgTime).toBeLessThan(PERFORMANCE_THRESHOLD)
      expect({
        name: 'Backstage passes',
        avgTime,
        withinThreshold: avgTime < PERFORMANCE_THRESHOLD,
      }).toMatchSnapshot({ avgTime: expect.any(Number) })
    })
  })

  describe('Conjured item performance', () => {
    it('should benchmark Conjured item performance', () => {
      const conjuredItem = new Item('Conjured', 1, 2)
      const gildedRose = new Shop([conjuredItem])

      const avgTime = benchmark(() => {
        gildedRose.updateQuality()
      })

      expect(avgTime).toBeLessThan(PERFORMANCE_THRESHOLD)
      expect({
        name: 'Conjured item',
        avgTime,
        withinThreshold: avgTime < PERFORMANCE_THRESHOLD,
      }).toMatchSnapshot({ avgTime: expect.any(Number) })
    })
  })

  describe('Mixed items performance', () => {
    it('should benchmark shop with multiple items', () => {
      const items = [
        new Item('+5 Dexterity Vest', 10, 20),
        new Item('Aged Brie', 2, 0),
        new Item('Sulfuras, Hand of Ragnaros', 0, 80),
        new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20),
        new Item('Conjured', 3, 6),
      ]
      const gildedRose = new Shop(items)

      const avgTime = benchmark(() => {
        gildedRose.updateQuality()
      })

      expect(avgTime).toBeLessThan(PERFORMANCE_THRESHOLD)
      expect({
        name: 'Mixed items',
        avgTime,
        withinThreshold: avgTime < PERFORMANCE_THRESHOLD,
      }).toMatchSnapshot({ avgTime: expect.any(Number) })
    })
  })
})
