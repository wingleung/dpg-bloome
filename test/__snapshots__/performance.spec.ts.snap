// Vitest Snapshot v1

exports[`Performance Tests > Aged Brie performance > should benchmark Aged Brie performance 1`] = `
{
  "avgTime": Any<Number>,
  "name": "Aged Brie",
  "withinThreshold": true,
}
`;

exports[`Performance Tests > Backstage passes performance > should benchmark Backstage passes performance 1`] = `
{
  "avgTime": Any<Number>,
  "name": "Backstage passes",
  "withinThreshold": true,
}
`;

exports[`Performance Tests > Conjured item performance > should benchmark Conjured item performance 1`] = `
{
  "avgTime": Any<Number>,
  "name": "Conjured item",
  "withinThreshold": true,
}
`;

exports[`Performance Tests > Mixed items performance > should benchmark shop with multiple items 1`] = `
{
  "avgTime": Any<Number>,
  "name": "Mixed items",
  "withinThreshold": true,
}
`;

exports[`Performance Tests > Sulfuras performance > should benchmark Sulfuras performance 1`] = `
{
  "avgTime": Any<Number>,
  "name": "Sulfuras",
  "withinThreshold": true,
}
`;

exports[`Performance Tests > default item > should benchmark default item performance 1`] = `
{
  "avgTime": Any<Number>,
  "name": "Default item",
  "withinThreshold": true,
}
`;

exports[`Performance Tests > empty state > should benchmark empty shop performance 1`] = `
{
  "avgTime": Any<Number>,
  "name": "Empty shop",
  "withinThreshold": true,
}
`;
